Python/Product/Debugger/DebugAdapterProcess.cs

```diff
@@ -15,41 +15,112 @@
 // permissions and limitations under the License.
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using Microsoft.PythonTools.Debugger.DebugEngine;
+using Microsoft.PythonTools.Infrastructure;
 using Microsoft.VisualStudio.Debugger.Interop.VSCodeDebuggerHost;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.PythonTools.Debugger {
     public sealed class DebugAdapterProcess : ITargetHostProcess {
-        private readonly Process _process;
-        private DebugAdapterProcess(ProcessStartInfo psi) {
+        private const int _debuggerConnectionTimeout = 5000; // 5000 ms
+        private const int _connectionCloseTimeout = 5000; // 5000 ms
+
+        private Process _process;
+        private readonly Guid _processGuid;
+        private PythonDebugOptions _debugOptions;
+        private string _interpreterOptions;
+        private int _listenerPort = -1;
+        private Stream _stream;
+        private Socket _socket;
+        private ManualResetEvent _connectedEvent = new ManualResetEvent(false);
+
+        public DebugAdapterProcess() {
+            _processGuid = Guid.NewGuid();
+        }
+
+        public static DebugAdapterProcess Start(string launchJson) {
+            var debugProcess = new DebugAdapterProcess();
+            debugProcess.StartProcess(launchJson);
+            return debugProcess;
+        }
+
+        private void StartProcess(string launchJson) {
+            InitializeListenerSocket();
+
+            var json = JObject.Parse(launchJson);
+            var exe = json["exe"].Value<string>();
+            var args = json["args"].Value<string>();
+            var cwd = json["cwd"].Value<string>();
+            ParseOptions(json["options"].Value<string>());
+
+            List<string> argsList = new List<string> {
+                string.IsNullOrWhiteSpace(_interpreterOptions) ? "" : _interpreterOptions,
+                PythonToolsInstallPath.GetFile("ptvsd_launcher.py").AddQuotes(),
+                cwd.Trim('\\').AddQuotes(),
+                _listenerPort.ToString(),
+                _processGuid.ToString(),
+                _debugOptions.ToString().AddQuotes(),
+                "-g",
+                args
+            };
+            var arguments = string.Join(" ", argsList);
+
+            ProcessStartInfo psi = new ProcessStartInfo {
+                FileName = exe,
+                Arguments = arguments,
+                WorkingDirectory = cwd,
+                RedirectStandardError = true,
+                RedirectStandardInput = false,
+                RedirectStandardOutput = false,
+                UseShellExecute = false,
+                CreateNoWindow = false,
+            };
+
+            var env = json["env"].Value<JArray>();
+            if (env.Count > 0) {
```

Comment:
This check shouldn't be necessary (empty collection is valid to iterate over). Is it meant to be a null check? Or was there something else inside the `if`?

Reply:
the check is not necessary.

---
In reply to: [153330672](https://github.com/Microsoft/PTVS/pull/3395#discussion_r153330672) [](ancestors = 153330672)